File: apps/ruby/design-pattern/05-strategy/lib/report.rb
# frozen_string_literal: true

module Strategy
  # Report class (Strategy pattern)
  #
  # Accepts a formatting strategy as a block or Proc,
  # and delegates output processing to it.
  class Report
    attr_reader :title, :text
    attr_accessor :formatter

    def initialize(&formatter)
      @title = "\u6708\u6b21\u5831\u544a"
      @text = %w[順調 最高の調子]
      @formatter = formatter
    end

    def output_report
      @formatter.call(self)
    end
  end
end

File: apps/scala/design-pattern/05-strategy/src/main/scala/Report.scala
type Formatter = (String, Seq[String]) => String

val htmlFormatter: Formatter = (title, text) =>
  val lines =
    Seq("<html>", " <head>", s"  <title>$title</title>", " </head>", "<body>") ++
      text.map(line => s"  <p>$line</p>") ++ Seq("</body>", "</html>")
  lines.mkString("\n")

val plainTextFormatter: Formatter = (title, text) =>
  (Seq(s"**** $title ****", "") ++ text).mkString("\n")

case class Report(
    title: String = "月次報告",
    text: Seq[String] = Seq("順調", "最高の調子"),
    formatter: Formatter = htmlFormatter
):
  def outputReport(): String = formatter(title, text)
  def withFormatter(formatter: Formatter): Report = copy(formatter = formatter)
